crates/chain-config/src/config/codec.rs

```diff
@@ -0,0 +1,151 @@
+mod decoder;
+mod encoder;
+mod parquet;
+
+pub use decoder::{
+    Decoder,
+    IntoIter,
+};
+pub use encoder::Encoder;
+
+use std::fmt::Debug;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Group<T> {
+    pub index: usize,
+    pub data: Vec<T>,
+}
+type GroupResult<T> = anyhow::Result<Group<T>>;
+
+#[cfg(test)]
+mod tests {
+    use std::ops::Range;
+
+    use crate::{
+        config::{
+            contract_balance::ContractBalance,
+            contract_state::ContractStateConfig,
+        },
+        CoinConfig,
+        ContractConfig,
+        MessageConfig,
+    };
+
+    use itertools::Itertools;
+
+    use super::*;
+
+    #[test]
+    fn writes_then_reads_written() {
+        let group_size = 100;
+        let num_groups = 10;
+        let starting_group_index = 3;
+        {
+            // Json
+            let temp_dir = tempfile::tempdir().unwrap();
+            let state_encoder = Encoder::json(temp_dir.path());
+
+            let init_decoder = || Decoder::json(temp_dir.path(), group_size).unwrap();
+
+            test_write_read(
+                state_encoder,
+                init_decoder,
+                group_size,
+                starting_group_index..num_groups,
+            );
+        }
+        {
+            // Parquet
+            let temp_dir = tempfile::tempdir().unwrap();
+            let state_encoder = Encoder::parquet(temp_dir.path(), 1).unwrap();
+
+            let init_decoder = || Decoder::parquet(temp_dir.path());
+
+            test_write_read(
+                state_encoder,
+                init_decoder,
+                group_size,
+                starting_group_index..num_groups,
+            );
+        }
+    }
+
+    fn test_write_read(
+        mut encoder: Encoder,
+        init_decoder: impl FnOnce() -> Decoder,
+        group_size: usize,
+        group_range: Range<usize>,
+    ) {
+        let num_groups = group_range.end;
+        let mut rng = rand::thread_rng();
```

Comment:
It's better to use seeded rng in tests:
https://docs.rs/rand/latest/rand/trait.SeedableRng.html

That way failures are reproducible. Or prop tests if you need to have coverage.

Reply:
Added it. The only implementations inside the `rand` crate had this warning:
> The algorithm is deterministic but should not be considered reproducible due to dependence on configuration and possible replacement in future library versions. For a secure reproducible generator, we recommend use of the [rand_chacha](https://crates.io/crates/rand_chacha) crate directly.

I'm fine with this level of reproducibility. If you deem the considerations above important enough to drag in another dependency please reopen this thread :)

Reply:
This is fine. I've used seedable many times and never had any trouble. It's just a precaution anyway.